pkg/controller/atlasmap/action_consolelink.go

```diff
@@ -0,0 +1,133 @@
+package atlasmap
+
+import (
+	"context"
+	routev1 "github.com/openshift/api/route/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/atlasmap/atlasmap-operator/pkg/apis/atlasmap/v1alpha1"
+	"github.com/atlasmap/atlasmap-operator/pkg/util"
+	"github.com/go-logr/logr"
+	consolev1 "github.com/openshift/api/console/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	consoleLinkFinalizer = "finalizer.console.openshift.io"
+)
+
+type consoleLinkAction struct {
+	baseAction
+}
+
+func newConsoleLinkAction(log logr.Logger, mgr manager.Manager) action {
+	return &consoleLinkAction{
+		newBaseAction(log, mgr, "ConsoleLink"),
+	}
+}
+func (action *consoleLinkAction) getRoute(ctx context.Context, atlasMap *v1alpha1.AtlasMap) (*routev1.Route, error) {
+	route := &routev1.Route{}
+	err := action.client.Get(ctx, types.NamespacedName{Name: atlasMap.Name, Namespace: atlasMap.Namespace}, route)
+	if err != nil && errors.IsNotFound(err) {
+		return route, nil
+	} else if err != nil {
+		action.log.Error(err, "Error retrieving route.")
+		return nil, err
+	}
+	return route, err
+}
+
+func (action *consoleLinkAction) handle(ctx context.Context, atlasMap *v1alpha1.AtlasMap) error {
+	isOpenShift, err := util.IsOpenShift(action.config)
+	if err != nil {
+		return err
+	}
+
+	if isOpenShift {
+
+		route, err := action.getRoute(ctx, atlasMap)
+		if err != nil {
+			return err
+		}
+
+		consoleLinkName := atlasMap.Name + "-" + atlasMap.Namespace
+		consoleLink := &consolev1.ConsoleLink{}
+		err = action.client.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+		if err != nil && errors.IsNotFound(err) {
+				consoleLink = createNamespaceDashboardLink(consoleLinkName, route, atlasMap)
+				err = action.client.Create(ctx, consoleLink)
+				if err != nil {
+					return err
+				} else if err == nil {
+					controllerutil.AddFinalizer(atlasMap, consoleLinkFinalizer)
```

Comment:
Just a question, does the removal of the `ConsoleLink` have to be done from a finalizer?

If you create the `ConsoleLink` with through `action.deployResource()`, it'll set a owner reference between the `AtlasMap` CR and the `ConsoleLink`.

Reply:
Console Link (application menu, namespace dashboard) is cluster-scoped. Namespaced dashboard which is part of the console link is namespace scoped. We can not set owner reference for console-link. I am using the finalizer as a flag to allow the deletion of console-link before the Atlasmap cr gets deleted and removes the flag once console-link has been deleted.

Reply:
Agree on cascading delete not working. However, we ended up doing this without a finalizer, by picking up the CR deletion reconcile call and cleaning it up in the reconcile loop the way we do adds and updates. See https://github.com/kiegroup/kie-cloud-operator/pull/397

Reply:
I removed the finalizer logic and handle in the reconciler. thank you for the suggestion.